### PR TITLE
Add gossip finished announcement, used by block proposer instead of fixed delay

### LINKED ISSUE #1453: Gossiper raise announcement consumed by block proposer
Add an announcement so that the block proposer knows when a deploy has been gossiped to sufficiently many peers and is safe to include in a proposed block.

https://casperlabs.atlassian.net/browse/NDRS-1140

### PR DESCRIPTION
**Pull request by Fraser999**
This PR adds a new `FinishedGossiping` announcement.

The `BlockProposer` has been updated to take advantage of this.  Rather than buffering a deploy when the `DeployAcceptor` announces it and waiting for a fixed duration before including that deploy in a block, the deploy is now only buffered when a `FinishedGossiping` announcement is made for it, and is immediately available for being included in a block.

This is built on top of #1475 and hence the first commit should not be reviewed as part of this PR.

Tagging @sacherjj @TomVasile @momipsl to draw attention to the change to the config file.

Closes #1453.

### CODE REVIEW

**Comment by goral09** [comment]
When is the `FinishedGossiping` announcement raised? Is it only when enough targets have been infected?

File: node/src/components/gossiper/gossip_table.rs [hunk 1]
```diff
@@ -443,40 +413,17 @@ impl<T: Copy + Eq + Hash + Display> GossipTable<T> {
...
-        if let Some(mut state) = self.current.remove(data_id) {
-            state.in_flight_count = 0;
-            let timeout = Instant::now() + self.finished_entry_duration;
-            let _ = self.paused.insert(*data_id, state);
-            let _ = self.paused_timeouts.push(timeout, *data_id);
+    /// We have deemed the data not suitable for gossiping further.  The entry will be marked as
+    /// `finished` and eventually be purged.
+    pub(crate) fn finish(&mut self, data_id: &T) {
+        if self.current.remove(data_id).is_some() {
+            self.insert_to_finished(data_id);
```
Comment (afck) [inline node/src/components/gossiper/gossip_table.rs:423]: And here we are not creating a `GossipAction::AnnounceFinished`?

File: node/src/components/gossiper/gossip_table.rs [hunk 2]
```diff
@@ -598,25 +529,11 @@ mod tests {
...
-        let action = gossip_table.new_partial_data(&data_id, node_ids[3]);
-        assert_eq!(GossipAction::AwaitingRemainder, action);
-        check_holders(&node_ids[..4], &gossip_table, &data_id);
-
         // Finish the gossip by reporting three infections, then check same partial data causes
         // `Noop` to be returned and holders cleared.
         let _ = gossip_table.new_complete_data(&data_id, Some(node_ids[0]));
         let limit = 4 + EXPECTED_DEFAULT_INFECTION_TARGET;
-        for node_id in &node_ids[4..limit] {
+        for node_id in &node_ids[3..limit] {
```
Comment (afck) [inline node/src/components/gossiper/gossip_table.rs:536]: Why do we need to infect one more now?

**Review by afck — APPROVED** [review]
I realized all my questions were about https://github.com/casper-network/casper-node/pull/1475.

The rest looks good to me. :+1:

File: node/src/components/gossiper.rs [hunk 3]
```diff
@@ -224,10 +232,6 @@ impl<T: Item + 'static, REv: ReactorEventT<T>> Gossiper<T, REv> {
         // in the entry being removed.
         if peers.is_empty() {
             self.metrics.times_ran_out_of_peers.inc();
-
-            self.table.pause(&item_id);
-            debug!(item=%item_id, "paused gossiping since no more peers to gossip to");
-            return Effects::new();
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply node/src/components/gossiper.rs:230]: No, we used to reduce `in_flight_count` to 0 inside `table.pause()`.

File: node/src/components/gossiper/gossip_table.rs (hunk 1 above)
Comment (Fraser999) [reply node/src/components/gossiper/gossip_table.rs:423]: I should probably have commented this, sorry!

We're only calling this method when the gossiper has failed to retrieve the deploy from storage.  That should never happen, but if it does, there's no point in emitting an `AnnounceFinished` since the `BlockProposer` will then also just try to get that deploy from storage.

Replying now though, it seems safer to return a `bool` here (`true` if the item was in `self.currrent`) and just let the gossiper announce it and have the `BlockProposer` fail too.

File: node/src/components/gossiper/gossip_table.rs (hunk 2 above)
Comment (Fraser999) [reply node/src/components/gossiper/gossip_table.rs:536]: We shouldn't - I'll update it and the other similar instance in the `new_complete_data()` test.

Comment (Fraser999) [reply node/src/components/gossiper/gossip_table.rs:536]: Fixed in [a3a385a](https://github.com/casper-network/casper-node/pull/1475/commits/a3a385ab94e1ee3fc31df04a17e9e5fdfda8f86f).

**Review by fizyk20 — APPROVED** [review]

**Comment by Fraser999** [comment]
@goral09 
> When is the `FinishedGossiping` announcement raised? Is it only when enough targets have been infected?

After the latest commits, it's after gossiping terminates for a given item for any reason:
* enough targets infected
* enough attempts to infect peers have been made
* no more peers available in the routing table
* the deploy is unavailable from storage (should never happen)

**Comment by goral09** [comment]
> @goral09
> 
> > When is the `FinishedGossiping` announcement raised? Is it only when enough targets have been infected?
> 
> After the latest commits, it's after gossiping terminates for a given item for any reason:
> 
> * enough targets infected
> * enough attempts to infect peers have been made
> * no more peers available in the routing table
> * the deploy is unavailable from storage (should never happen)

Thanks. I'm wondering whether under certain conditions this won't lead to deploy starvation? Or a situation where we should propose a block and we have a deploy but we create empty block b/c none of these conditions has been met?

File: node/src/components/gossiper/gossip_table.rs (hunk 1 above)
Comment (Fraser999) [reply node/src/components/gossiper/gossip_table.rs:429]: Fixed in [b1e63fe](https://github.com/casper-network/casper-node/pull/1484/commits/b1e63fe46556005af8f4a504cdfe5fb8dd77efeb).

**Comment by Fraser999** [comment]
I'm expecting that the large-scale network tests will show up if we're actually pessimising here.  My local nctl tests are showing good results in terms of how quickly deploys get included in blocks now, but I haven't validated that every deploy was actually included in a some block.

I expect that whichever of the gossiping termination conditions fires for a given deploy, it will be reasonably fast.  I can't see how any deploy could be left in the `current` state for any significant length of time, and if that _is_ possible, then it's a bug which can cause a resource issue, since we only purge the `finished` items from the gossiper.

> Do you think that, in practice, enough attempts to infect peers have been made should be quick enough to fire?

Well, it depends upon the timing of the deploy arriving at the node vs the block proposer being asked to propose a new block.  But ultimately, this new approach for delaying deploy inclusion will likely be better than the current delay of 1 minute for every deploy.  Most or all deploys will be delayed for significantly less time I expect.

The unknown for now is whether our expectation was correct: that this should give enough time for the deploy to be disseminated so that peers don't have to request the deploy when they learn of a newly-proposed block.  Again, I hope this will be proved correct when we run the large scale network test.

**Comment by goral09** [comment]
> I'm expecting that the large-scale network tests will show up if we're actually pessimising here. My local nctl tests are showing good results in terms of how quickly deploys get included in blocks now, but I haven't validated that every deploy was actually included in a some block.

They aren't designed to smoke out any edge cases since the whole network is comprised of fairly beefy machines with good network connections.
 
> I expect that whichever of the gossiping termination conditions fires for a given deploy, it will be reasonably fast. I can't see how any deploy could be left in the `current` state for any significant length of time, and if that _is_ possible, then it's a bug which can cause a resource issue, since we only purge the `finished` items from the gossiper.

What about a case where packets are lost or a peer does not respond?

> > Do you think that, in practice, enough attempts to infect peers have been made should be quick enough to fire?
> 
> Well, it depends upon the timing of the deploy arriving at the node vs the block proposer being asked to propose a new block. But ultimately, this new approach for delaying deploy inclusion will likely be better than the current delay of 1 minute for every deploy. Most or all deploys will be delayed for significantly less time I expect.
> 
> The unknown for now is whether our expectation was correct: that this should give enough time for the deploy to be disseminated so that peers don't have to request the deploy when they learn of a newly-proposed block. Again, I hope this will be proved correct when we run the large scale network test.

I would think that disseminating a deploy to enough peers before proposing it is a secondary requirement to proposing a deploy early. Ideally, we would solve both but I think we should not be holding up transaction in the mempool b/c we haven't gossiped it to enough peers. What do you think about adding a fallback mechanism that a deploy will be proposed in the next _next_ round even if we hadn't infected enough peers – i.e. if a deploy arrives before the next block we might not include it if the we hadn't infected enough peers but if there is second proposal after that then we would not hold it up any longer?

**Comment by Fraser999** [comment]
> They aren't designed to smoke out any edge cases since the whole network is comprised of fairly beefy machines with good network connections.

Yes, I wasn't really meaning for this to be a smoke test.  Rather, that if this change results in a large increase in traffic, we'll detect it there.

> What about a case where packets are lost or a peer does not respond?

There is a configurable timeout for every gossip request sent.  If a peer doesn't respond, it won't stop the item eventually being moved to finished and the gossiper announcement being made.

> I would think that disseminating a deploy to enough peers before proposing it is a secondary requirement to proposing a deploy early. Ideally, we would solve both but I think we should not be holding up transaction in the mempool b/c we haven't gossiped it to enough peers.

But we're already holding up every deploy/transaction for 1 minute before proposing it.  We added this in #1283 with the express intent of allowing the deploys to be disseminated before proposal.  This PR is intended to reduce that delay in most cases.

**Comment by afck** [comment]
> a fallback mechanism that a deploy will be proposed in the next next round even if we hadn't infected enough peers

That sounds dangerous to me. If gossiping works fine this will never be necessary. If for some reason a deploy fails to get gossiped, I'd rather not risk including it in a block: The same problem that held up the deploy would now stall consensus.

**Comment by goral09** [comment]
> > a fallback mechanism that a deploy will be proposed in the next next round even if we hadn't infected enough peers
> 
> That sounds dangerous to me. If gossiping works fine this will never be necessary. If for some reason a deploy fails to get gossiped, I'd rather not risk including it in a block: The same problem that held up the deploy would now stall consensus.

I don't think so though. The failure to raise `FinalizationAnnouncement` can be caused by many things (a bug in the gossiper, in the logic detecting whether to raise the announcement, peers not responding to events, etc.), I don't think there's a reason to believe that it would also stall consensus. 


Anyway, Fraser seems to be convinced this is not going to be a problem and I don't have any evidence it will be so, I am fine with this.

**Comment by goral09** [comment]
> But we're already holding up every deploy/transaction for 1 minute before proposing it. We added this in #1283 with the express intent of allowing the deploys to be disseminated before proposal. This PR is intended to reduce that delay in most cases.

This PR though removes that code so the timeout-based announcement is never going to be raised. My point was that this approach seems, to me, more robust in the "good case" but also more dangerous in the "bad case".

File: node/src/components/gossiper/gossip_table.rs (hunk 1 above)
Comment (fizyk20) [reply node/src/components/gossiper/gossip_table.rs:429]: I'd actually prefer to treat this case differently and have `AnnounceFinished` mean a _successful_ end of gossiping. We could emit another announcement here, if we want to handle it somewhere else. It seems a bit awkward to me that we'll signal to the block proposer something to the effect of "hey, you can now propose this" when we know that we don't even have this piece of data :grimacing:

Comment (Fraser999) [reply node/src/components/gossiper/gossip_table.rs:429]: As discussed offline, I've just added a comment to help clarify things should we want to take a different approach here in the future.  Added in [c8c8c3b](https://github.com/casper-network/casper-node/pull/1484/commits/c8c8c3b61502f81a15bdad55ece0de98ae43d184).

**Comment by Fraser999** [comment]
bors merge
